076-decouple-ca-management-from-end-entity-certificate-management.md

```diff
@@ -0,0 +1,389 @@
+   - This means that end-entity certificate generation and the rollout of trust in the CA certificates referenced in those end-entity certificates, needs to be decoupled.
+ - These solutions might take a varying amount of time to issue a certificate.
+   - Since we do not know how long certificate issuance might take, we can't block the reconcile loop waiting for certificates to be issued.
+   - This means it can't be assumed that e.g. the CA certificate is already present when the component reconcilers run (since it not be issued yet).
+This would be highly valuable for organizations with compliance requirements with regard to certificates.
+
+Currently, it is hard to integrate one of these certificate management solutions for a couple of reasons:
+ - The state of certificates, particularly CA certificates is determined based on multiple different annotations and fields on internal Java objects.
+   This makes it hard to comprehend the current state, particularly when doing a renewal and tricky to reason about edge cases that might occur.
+ - The component reconcilers directly interact with the Java class that represents the Cluster CA.
+   For example, this class contains methods not only to manage the CA certificate and private key, but also generate certificates for Strimzi components (like Cruise Control) and store state about renewals that happened during the current reconcile loop.
+   This makes it hard to drop in an alternative to the existing Strimzi Cluster CA, since it's behaviour is relied upon throughout the reconcile loop.
+
+This proposal describes changes to make it easier to reason about the current state of CA certificates and track the roll out of trust.
+
+## Proposal
+
+Strimzi should be updated to decouple the management of the cluster CA certificate, and the consumption of the CA certificate by components (such as ZooKeeper, Kafka and cluster operator).
+
+To do this we will introduce two new Kubernetes Secrets.
+The first Kubernetes Secret, called <CLUSTER_NAME>-cluster-ca-trust-state (referred to from here onwards as trust-state Secret) is used to keep track of all Cluster CA certificates that are currently in use and their current state with respect to issued end entity certificates.
+The second Kubernetes Secret, called <CLUSTER_NAME>-cluster-ca-trusted-certs (referred to from here onwards as trusted-certs Secret) is used to store all Cluster CA certificates that are currently trusted by components.
+The trusted-certs Secret is mounted onto Kubernetes Pods and used for their TrustStore.
+These new Kubernetes Secrets are separate from the ones that are used by Strimzi to store the latest CA certificate and key (or by the user when they are managing their own CA certificates).
+The cluster operator will update these new Kubernetes Secrets when it renews the cluster CA certificate (or observes that the user has updated them).
+
+Introducing these new Kubernetes Secrets brings a couple of benefits:
```

Comment:
While the _trusted_ secret makes sense internally, ou also need to deal with how users will trust the certificates. If we cannot trust the CA certificate Secret, so can't they. So this secret will need to be exposed and that makes this a bit complicated because it would be a breaking change?

Reply:
That's true, this would represent a breaking change for application developers who are currently mounting the `<CLUSTER_NAME>-cluster-ca-cert` Secret. Do you think that would be a blocker to this being accepted as a proposal?